lib/apipie_bindings/authenticators/negotiate.rb

```diff
@@ -0,0 +1,66 @@
+require 'apipie_bindings/authenticators/base'
+
+module ApipieBindings
+  module Authenticators
+    # Negotiate authenticator
+    # Implements gssapi negotiation with preexisting kerberos ticket
+    # Requires a authentication url, the authentication request will be against.
+    # This url needs to support auth negotiation and after successful auth it should return 'set-cookie' header with session.
+    # This session will be initiated in the auth request and the original request will be made with this cookie.
+    # Next requests should be already skip the negotiation, please implement Session support in your client, for not using the negotiation on every request.
+    class Negotiate < Base
+
+      # Creates new authenticator for Negotiate auth
+      # @param [String] url to make authentication request to.
+      # @param [Hash] auth_request_options passed to RestClient::Request - especially for SSL options
+      #   see https://github.com/rest-client/rest-client/blob/master/lib/restclient/request.rb.
+      # @option service service principal used for gssapi tickets - defaults to HTTP.
+      # @option method http method used for the auth request - defaults to 'get'.
+      def initialize(authorization_url, auth_request_options = {})
+        @authorization_url = authorization_url
+        @service = auth_request_options.delete(:service) || 'HTTP'
+        auth_request_options[:method] ||= 'get'
+        @auth_request_options = auth_request_options
+      end
+
+      def error(ex)
+        if ex.is_a?(GSSAPI::GssApiError)
+          raise ApipieBindings::AuthenticatorError.new(:negotiate, :no_context, ex)
+        elsif ex.is_a?(ApipieBindings::ConfigurationError)
+          raise ApipieBindings::AuthenticatorError.new(:negotiate, :configuration, ex)
+        else
+          raise ex
+        end
+      end
+
+      def authenticate(original_request, args)
+        require 'gssapi'
+        uri = URI.parse(@authorization_url)
+        @gsscli = GSSAPI::Simple.new(uri.host, @service)
+
+        token = @gsscli.init_context
+        headers = { 'Authorization' => "Negotiate #{Base64.strict_encode64(token)}" }
+
+        RestClient::Request.execute(@auth_request_options.merge(headers: headers, url: @authorization_url)) do |response, request, raw_response|
+          if response.code == 401
+            raise RestClient::Unauthorized.new(response), 'Negotiation authentication did not pass.'
+          end
+
+          if response.code == 302
```

Comment:
But wait, there's a twist...

At least with Foreman, when extlogin fails it redirects to `/users/login`, however if it passes, it redirects too, although to `/hosts`.

```
# hammer log
[ERROR 2022-05-05T11:08:59 Exception] Server misconfiguration detected
  - have you run installer with option --foreman-ipa-authentication=true?

Server misconfiguration detected
  - have you run installer with option --foreman-ipa-authentication=true?
[ERROR 2022-05-05T11:08:59 Exception]

ApipieBindings::ConfigurationError (Server misconfiguration detected):
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/authenticators/negotiate.rb:50:in `block in authenticate'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:807:in `process_result'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:725:in `block in transmit'
    /usr/share/ruby/net/http.rb:933:in `start'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:715:in `transmit'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:145:in `execute'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:52:in `execute'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/authenticators/negotiate.rb:44:in `authenticate'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/api/session_authenticator_wrapper.rb:53:in `authenticate'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/rest_client_extensions.rb:14:in `block in <module:RestClientExtensions>'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:706:in `block in transmit'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:705:in `each'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:705:in `transmit'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:145:in `execute'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/request.rb:52:in `execute'
    /usr/share/gems/gems/rest-client-2.0.2/lib/restclient/resource.rb:51:in `get'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:327:in `call_client'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:240:in `http_call'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:190:in `call_action'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:185:in `call'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/action.rb:14:in `call'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/api/connection.rb:34:in `login'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/authenticate/login.rb:12:in `execute_with_params'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/auth.rb:38:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/bin/hammer:147:in `<top (required)>'
    /usr/bin/hammer:23:in `load'
    /usr/bin/hammer:23:in `<main>'
```

```
# production.log
2022-05-05T11:08:53 [I|app|311a2aec] Started GET "/users/extlogin" for 192.168.122.247 at 2022-05-05 11:08:53 +0000
2022-05-05T11:08:53 [I|app|311a2aec] Processing by UsersController#extlogin as */*
2022-05-05T11:08:53 [I|app|311a2aec] Authorized user foo(f f)
2022-05-05T11:08:59 [I|app|311a2aec] User 'foo' logged in from '192.168.122.247'
2022-05-05T11:08:59 [I|app|311a2aec] Redirected to https://april-nightingale.example.com/hosts
2022-05-05T11:08:59 [I|app|311a2aec] Completed 302 Found in 5774ms (ActiveRecord: 29.3ms | Allocations: 18997)
```